src/librustc/infer/error_reporting.rs

```diff
@@ -555,6 +555,155 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
+    fn highlight_outer(&self,
+                       value: &mut Vec<(String, bool)>,
+                       other_value: &mut Vec<(String, bool)>,
+                       name: String,
+                       sub: &ty::subst::Substs<'tcx>,
+                       pos: usize,
+                       other_ty: &ty::Ty<'tcx>) {
+        value.push((name, true));
+        let len = sub.len();
+        if len > 0 {
+            value.push(("<".to_string(), true));
+        }
+
+        let sts = sub.regions();
+        for (i, st) in sts.enumerate() {
+            value.push((format!("{}", st), false));
+
+            if len > 0 && i != len - 1 {
+                value.push((format!(", "), false));
+            }
+        }
+
+        let sts = sub.types();
+        for (i, st) in sts.enumerate() {
+            if i == pos {
+                let (v, o_v) = self.cmp(st, other_ty);
+                value.extend(v);
+                other_value.extend(o_v);
+            } else {
+                value.push((format!("{}", st), true));
+            }
+
+            if len > 0 && i != len - 1 {
+                value.push((format!(", "), true));
+            }
+        }
+        if len > 0 {
+            value.push((">".to_string(), true));
+        }
+    }
+
+
+    fn cmp(&self, t1: ty::Ty<'tcx>, t2: ty::Ty<'tcx>)
+        -> (Vec<(String, bool)>, Vec<(String, bool)>)
+    {
+        match (&t1.sty, &t2.sty) {
+            (&ty::TyAdt(def1, sub1), &ty::TyAdt(def2, sub2)) => {
+                let mut values: (Vec<(String, bool)>, Vec<(String, bool)>) = (vec![], vec![]);
+                let name1 = self.tcx.item_path_str(def1.did.clone());
+                let name2 = self.tcx.item_path_str(def2.did.clone());
+                if name1 == name2 {
```

Comment:
when you say "subtype", are you referring to the `X` in `Option<X>`? If so, subtype isn't quite the right term (that usually is used to mean e.g. the relationship between `String` and `Object` in Java). I would say "type argument", I guess, if I were referring to the `X` in some use of `Option`, and "type parameter" if I am referring to the *definition* of `Option`.